workflow/controller/scope.go

```diff
@@ -57,26 +60,71 @@ func (s *wfScope) resolveVar(v string) (interface{}, error) {
 	return nil, errors.Errorf(errors.CodeBadRequest, "Unable to resolve: {{%s}}", v)
 }
 
-func (s *wfScope) resolveParameter(v string) (string, error) {
-	val, err := s.resolveVar(v)
+func (s *wfScope) resolveParameter(p *wfv1.ValueFrom) (string, error) {
+	if p == nil {
+		return "", nil
+	}
+	if p.Parameter == "" && p.Expression == "" {
+		return "", nil
+	}
+	var val interface{}
+	var err error
+	if p.Expression != "" {
+		val, err = expr.Eval(p.Expression, s.getAllParamArtifact())
+	} else {
+		val, err = s.resolveVar(p.Parameter)
+	}
+
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("unable to resolve expression: %s", err)
 	}
-	valStr, ok := val.(string)
-	if !ok {
-		return "", errors.Errorf(errors.CodeBadRequest, "Variable {{%s}} is not a string", v)
+	if val == nil {
+		return "", nil
 	}
-	return valStr, nil
+	return fmt.Sprintf("%v", val), nil
 }
 
-func (s *wfScope) resolveArtifact(v string, subPath string) (*wfv1.Artifact, error) {
-	val, err := s.resolveVar(v)
+func (s *wfScope) getAllParamArtifact() map[string]interface{} {
```

Comment:
I don't think we should need this func, can you use `resolveParameter` and `resolveArtifact` instead?